projects/cli-files/go-cat/cmd/root.go

```diff
@@ -0,0 +1,59 @@
+package cmd
+
+import (
+	"bufio"
+	"fmt"
+	"os"
+)
+
+func Execute() {
+
+	// if there is no argument given then error
+	if len(os.Args) < 2 {
+		fmt.Fprintf(os.Stderr, "no filename provided\n")
+		os.Exit(2)
+	}
+
+	// if there is an argument provided then save that argument as the filename
+	filename := os.Args[1]
+
+	// print out the filename to check
+	// fmt.Println("--- DEBUG filename", filename)
+
+	// check if the file exists
+	// fileInfo, err := os.Stat(filename)
+	_, err := os.Stat(filename)
+
+	// if the file does not exist then error
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "no such file or directory: %s\n", filename)
+		os.Exit(1)
+	}
+
+	// print out the fileInfo to check
+	// fmt.Println("--- DEBUG fileInfo", fileInfo)
+
+	// attempt to open the file
+	file, err := os.Open(filename)
+
+	// if the file cannot be opened then error
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "cannot open the file: %v\n", err)
+		os.Exit(1)
+	}
+
+	// print out the file to check
+	// fmt.Println("--- DEBUG file", file)
+
+	// make sure to close the file at the end of the function
+	defer file.Close()
+
+	// make a scanner to read from the file
+	scanner := bufio.NewScanner(file)
+
+	// use the scanner to scan through the file, line by line, printing out each line
+	for scanner.Scan() {
+		fmt.Fprint(os.Stdout, scanner.Text(), "\n")
+	}
```

Comment:
What happens if scanner is interrupted (for any given reason), do we need to inform the user, perhaps by exit codes 0 or 1?

What about using `os.ReadFile`? I'm not saying it's the right way, but what are the pros/cons of using `ReadFile` vs using a scanner?

Reply:
from what i read about this

os.ReadFile loads the entire file into memory
which depending on how big that file is can be inefficient
in my head i am thinking of some massive csv file with millions of entries
but on a small file this is better
gives an easy error back

Scanner loads one line (token(?)) at a time (incrementally line by line)
so it won't take up a lot of memory
and you can customise the Scanner to change what it sees as a "line"
have to handle errors manually

maybe the intention of this project is to use only `os` packages
so for the small .txt files in assets os.ReadFile is better

I don't know what happens with a scanner interrupt i will look into it

Reply:
Yeah your thought process is right, scanner is definitely a better choice for handling large files. But there's a chance that we don't return the full file contents to the user due to interruptions (in this particular project you might not expect interruptions, but we should still make considerations for edge cases). 

`scanner.Scan()` returns multiple values, so you could potentially use those values to deal with interruptions. But I'd only consider that as an extesnsion to the project.

In terms of this project, either solutions would work but without handling the output from `scanner.Scan()` the solution would be _incomplete_